esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -350,13 +350,35 @@ void EsdlServiceImpl::addMethodLevelRequestTransform(const char *method, IProper
     }
 }
 
-static void ensureMergeOrderedPTree(Owned<IPropertyTree> &target, IPropertyTree *src)
+static void ensureMergeOrderedEsdlTransform(Owned<IPropertyTree> &dest, IPropertyTree *src)
 {
     if (!src)
         return;
-    if (!target)
-        target.setown(createPTree(ipt_ordered));
-    mergePTree(target, src);
+    if (!dest)
+        dest.setown(createPTree(ipt_ordered));
+    //copy so we can make changes, like adding calculated targets
+    Owned<IPropertyTree> copy = createPTreeFromIPT(src, ipt_ordered);
+    const char *target = copy->queryProp("@target");
+    if (target && *target)
+    {
+        Owned<IPropertyTreeIterator> children = copy->getElements("*");
+        ForEach(*children)
+        {
+            const char *existing = children->query().queryProp("@target");
+            if (existing && *existing)
+            {
+                if (*existing=='.') //support in future so error now so we don't have backward compatability issues
+                {
+                    VStringBuffer errorXpath("unsupported choose xpath*>>%s", existing); //will cause request time xpath error
+                    children->query().setProp("@target", errorXpath);
+                    DBGLOG("ESDL Binding: request transform choose xpath error: %s", existing);
+                }
+                continue;
+            }
+            children->query().setProp("@target", target);
```

Comment:
Is this the best name for the attribute?

While not needed now, I can see a scenario where we would extend this mechanism to the transform rules themselves. I have use cases where the transform will set an in-band and an out-of-band value based on the same test result. By only setting targets on the `choose` elements it is necessary to replicate, and repeat the execution of, the test logic for each value set.

Maybe `root` or `base` would be something that makes sense now and could eventually be added to the rules that already include `target` attributes.

Reply:
I thought about something like “baseTarget” but my current idea is to later support both relative and absolute targets.  A relative target is relative to the parent target.  An absolute target is self contained.  That’s why i error when the path starts with “.” Because in the future that would mean relative.

But we could do the base target thing if we don’t think support for relative paths is enough.

Reply:
The rule's `target` is always relative to whatever root node is used in the evaluation. I'm not questioning how this root is selected. I'm just questioning the name of the attribute used to select it, and only because I can see that attribute being added to the rule some day, and two `target` attributes won't work.

I like the idea of supporting relative path references for the attribute, whatever it may be named.

Reply:
I'm changing this to an attribute intended only for internal use.  CRT/@target is added to the xsdl:choose as xsdl:choose/@_crtTarget.  But not meant to be overridden by the user at all.  We can add support for use defined relative targets at different levels later.  Main reason is I want to get the relative path syntax exactly right.